### PR TITLE
handle flash slow tests

### PR DESCRIPTION
**Pull request by itazap**
fixing broken slow / flash tests as we cannot pass the regular 4dim attention mask for flash

_update_causal_mask does not work with the patch cross attention , handling patch attn mask update for flash in the _prepare function

### CODE REVIEW

**Comment by itazap** [comment]
run-slow: blt

**Comment by ydshieh** [comment]
Thank you @itazap . I will let @vasqu or @Cyrilvallez or @ArthurZucker to review this.

Searching `"flash" in self.config._attn_implementation` within the codebase, I can see this happens mostly within methods `def _update_xxx_mask`, like `def _update_cross_attn_mask`, `_update_causal_mask`, etc.

So probably it's better to do the same for this model, but I'm not super sure if this values a lot.

**Comment by ydshieh** [comment]
And thank you for use `run-slow`.

It turns out that your github username is not in `.github/workflows/self-comment-ci.yml`, so the CI is not triggered.

I will open a PR to add you, then you can rebase on main and use `run-slow` again 🙏

**Comment by ydshieh** [comment]
Could you approve

https://github.com/huggingface/transformers/pull/41094

🙏

**Comment by ydshieh** [comment]
Merged #41094. You can rebase and try `run-slow` again.

(or if you already try to run the tests with a SSH runner, including slow tests , it's fine)

File: src/transformers/models/blt/modular_blt.py [hunk 1]
```diff
@@ -356,6 +356,9 @@ def forward(
         if self.config._attn_implementation != "eager":
             attention_interface = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
 
+        if "flash" in self.config._attn_implementation:
+            attention_mask = None
```
Comment (Cyrilvallez) [inline src/transformers/models/blt/modular_blt.py]: Well, if we have any padding or similar, this is wrong here

**Review by Cyrilvallez — COMMENTED** [review]
We cannot simply skip the mask! It needs to account for padding!

**Comment by itazap** [comment]
run-slow: blt

File: src/transformers/models/blt/modular_blt.py (hunk 1 above)
Comment (itazap) [reply src/transformers/models/blt/modular_blt.py]: I added converting the patch cross attn mask!

File: tests/models/blt/test_modeling_blt.py [hunk 2]
```diff
@@ -284,6 +284,18 @@ def test_model_rope_scaling_from_config(self, scaling_type):
...
+    def test_flash_attn_2_equivalence(self):
+        pass
+
+    @unittest.skip(reason="BLT's patch cross-attention results in different outputs with flash attention")
+    def test_flash_attn_2_inference_equivalence(self):
+        pass
+
+    @unittest.skip(reason="BLT's patch cross-attention results in different outputs with flash attention")
+    def test_flash_attn_2_inference_equivalence_right_padding(self):
+        pass
```
Comment (vasqu) [inline tests/models/blt/test_modeling_blt.py]: Does it still not work with the current version? Imo, maybe we should just set `_supports_flash_attn=False`? 

This seems a bit sus tbh as fa is flaky but not consistently failing usually.

File: src/transformers/models/blt/modular_blt.py [hunk 3]
```diff
@@ -866,7 +879,13 @@ def forward(
...
 
         cross_attn_mask_enc = _prepare_patch_cross_attention_mask(
-            patch_ids, patch_lengths.shape[1], sequence_length, True, self.config.cross_attn_k, encoder_embeds.dtype
+            patch_ids,
+            patch_lengths.shape[1],
+            sequence_length,
+            True,
+            self.config.cross_attn_k,
+            encoder_embeds.dtype,
+            self.config._attn_implementation,
```
Comment (vasqu) [inline src/transformers/models/blt/modular_blt.py]: kwargs would help a lot on these to have a quick first glance

File: src/transformers/models/blt/modular_blt.py [hunk 4]
```diff
@@ -160,10 +161,11 @@ def _prepare_patch_cross_attention_mask(
         patches_as_queries (bool): If True, patches are used as queries, otherwise as keys.
         cross_attn_k (int): Cross-attention multiplier for repeating patches.
         dtype (torch.dtype): Data type for the output mask.
+        attn_implementation (str): Attention implementation type to determine mask format.
 
     Returns:
         Tuple[torch.Tensor, torch.Tensor]:
-            - cross_attention_mask: 4D tensor [batch_size, 1, q_len, kv_len]
+            - cross_attention_mask: 4D tensor [batch_size, 1, q_len, kv_len] or 2D for flash attention
```
Comment (vasqu) [inline src/transformers/models/blt/modular_blt.py]: Let's type the proper shape then, e.g. `bsz, kv_len`

File: src/transformers/models/blt/modular_blt.py [hunk 5]
```diff
@@ -204,17 +206,28 @@ def _prepare_patch_cross_attention_mask(
...
-    # Invert the mask (following mllama pattern exactly)
-    # True -> 0.0 (attend), False -> 1.0 (will become -inf)
-    inverted_cross_attn_mask = 1.0 - cross_attention_mask.to(dtype)
-    cross_attention_mask = inverted_cross_attn_mask.masked_fill(
-        inverted_cross_attn_mask.to(torch.bool), torch.finfo(dtype).min
-    )
+        # flash attention expects 2D mask [batch, seq_len]
+        # so if we have a 4D mask [batch, 1, q_len, kv_len], we need to convert it
+        if flash_mask.dim() == 4:
+            flash_mask = flash_mask.squeeze(1).any(dim=1).to(dtype)
```
Comment (vasqu) [inline src/transformers/models/blt/modular_blt.py]: This is a bit weird, we unsqueeze to 4D either way, no? So this is always 4D

Also not sure if we should check on q_len rather than kv_len here since the cross attentions are based on the "encoder" mask.

**Review by vasqu — COMMENTED** [review]
Not entirely sure if this working as expected especially since we skip the essential fa tests that would check this for us.

Left some comments, maybe we can get it fixed or mark it as not supported for now?

**Comment by itazap** [comment]
run-slow: blt

**Comment by itazap** [comment]
run-slow: blt

File: tests/models/blt/test_modeling_blt.py (hunk 2 above)
Comment (itazap) [reply tests/models/blt/test_modeling_blt.py]: hey @vasqu thanks for the review! I updated the tolerances but still getting failures on some tests that I honestly don't know where to address in terms of the modeling code, so I guess we can remove flash support unless you might have some advice on addressing them? thanks in advance! 
https://github.com/huggingface/transformers/actions/runs/18008591132/job/51234995082#:~:text=Set%20TORCHDYNAMO_VERBOSE=,exit%20code%201.

Comment (vasqu) [reply tests/models/blt/test_modeling_blt.py]: Nope, nothing concrete in mind :D might be that the cross mask is a bit more complex? Would need to visualize it and look what's going on there.

Would leave it for now and mark as unsupported to get most tests at least running again.

**Comment by itazap** [comment]
run-slow: blt

File: src/transformers/models/blt/modeling_blt.py [hunk 6]
```diff
@@ -955,6 +955,7 @@ def _prepare_patch_cross_attention_mask(
     patches_as_queries: bool = False,
     cross_attn_k: int = 1,
     dtype: torch.dtype = torch.float32,
+    attn_implementation: str = "eager",
```
Comment (vasqu) [inline src/transformers/models/blt/modeling_blt.py]: We don't need this anymore since eager / sdpa take similar masks

File: src/transformers/models/blt/modeling_blt.py [hunk 7]
```diff
@@ -1148,12 +1156,13 @@ def forward(
...
-            False,
-            self.config.cross_attn_k,
-            encoder_embeds.dtype,
+            patch_ids=decoder_patch_ids,
+            num_patches=patch_lengths.shape[1],
+            sequence_length=sequence_length,
+            patches_as_queries=False,
+            cross_attn_k=self.config.cross_attn_k,
+            dtype=encoder_embeds.dtype,
+            attn_implementation=self.config._attn_implementation,
```
Comment (vasqu) [inline src/transformers/models/blt/modeling_blt.py]: :heart:

File: tests/models/blt/test_modeling_blt.py [hunk 8]
```diff
@@ -245,6 +245,14 @@ def test_eager_matches_sdpa_inference(
             self, name, torch_dtype, padding_side, use_attention_mask, output_attentions, enable_kernels, atols=atols
         )
 
+    @unittest.skip("BLT does not support flash attention")
+    def test_flash_attn_2_can_compile_with_attention_mask_None_without_graph_break(self):
+        pass
```
Comment (vasqu) [inline tests/models/blt/test_modeling_blt.py]: Shouldnt need a skip since it should require fa installed + supported, if not we need to add a decorator here that we missed

File: tests/models/blt/test_modeling_blt.py [hunk 9]
```diff
@@ -245,6 +245,14 @@ def test_eager_matches_sdpa_inference(
...
+
+    @unittest.skip("BLT does not support flash attention dispatch")
+    def test_sdpa_can_dispatch_on_flash(self):
+        pass
```
Comment (vasqu) [inline tests/models/blt/test_modeling_blt.py]: I think you always create a mask regardless what happens?

This only works if sdpa gets no mask, so let's fix the reason to something along "Always applies a mask so SDPA cannot dispatch to flash attention"

**Review by vasqu — APPROVED** [review]
Thanks, my last few nits :hugs: 

The failing test is an integration test, not really related to this PR but should be fixed at some point

**Comment by itazap** [comment]
run-slow: blt

File: tests/models/blt/test_modeling_blt.py (hunk 8 above)
Comment (itazap) [reply tests/models/blt/test_modeling_blt.py]: agreed but it does indeed run without the skip, you mean we already have a decorator I can add here?

**Comment by itazap** [comment]
run-slow: blt
